### PR TITLE
324-add-tensorrt-support

### LINKED ISSUE #324: Add TensorRT support for a part of bundles
@binliunls has successfully enable TensorRT for some of the bundles offline.
This ticket is used to track the progress of updating it into the repository.

The following parts will be addressed respectively:

- [x] update `spleen_ct_segmentation`, `endoscopic_tool_segmentation` and `pathology_tumor_detection` with TensorRT support. The readme should also contain benckmarks. #325 
- [x] enable export TensorRT CI test
    - [x] It requires API support. Related ticket in MONAI core is: https://github.com/Project-MONAI/MONAI/pull/5986
    - [x] #326 
    - [x] It requires a blossom-ci pipeline to trigger the test script weekly.

### PR DESCRIPTION
**Pull request by binliunls**
Fixes #324 .

### Description
Add the command line of TensorRT convert to a part of bundles. Update the TensorRT benchmark results to these bundles.

### Status
**Ready/Work in progress/Hold**

### Please ensure all the checkboxes:

- [x] Codeformat tests passed locally by running `./runtests.sh --codeformat`.
- [ ] In-line docstrings updated.
- [ ] Update `version` and `changelog` in `metadata.json` if changing an existing bundle.
- [ ] Please ensure the naming rules in config files meet our requirements (please refer to: `CONTRIBUTING.md`).
- [ ] Ensure versions of packages such as `monai`, `pytorch` and `numpy` are correct in `metadata.json`.
- [ ] Descriptions should be consistent with the content, such as `eval_metrics` of the provided weights and TorchScript modules.
- [ ] Files larger than 25MB are excluded and replaced by providing download links in `large_file.yml`.
- [ ] Avoid using path that contains personal information within config files (such as use `/home/your_name/` for `"bundle_root"`).

### CODE REVIEW

File: models/endoscopic_tool_segmentation/docs/README.md
Comment (Nic-Ma) [inline models/endoscopic_tool_segmentation/docs/README.md]: I think no need to put 2 commands here, maybe just 1 command and explain that we can change the precision to FP16 by `--precision fp16`.

What do you think?

Thanks.

File: models/endoscopic_tool_segmentation/configs/metadata.json
Comment (yiheng-wang-nv) [inline models/endoscopic_tool_segmentation/configs/metadata.json]: Mark: may need to modify `monai_version` after 1.2 is released.

File: models/endoscopic_tool_segmentation/docs/README.md
Comment (binliunls) [reply models/endoscopic_tool_segmentation/docs/README.md]: Thank you for your advice. I have updated this part.

File: models/endoscopic_tool_segmentation/configs/metadata.json
Comment (Nic-Ma) [reply models/endoscopic_tool_segmentation/configs/metadata.json]: Hi @yiheng-wang-nv ,

According to our plan, please create a `monai_1.2` branch in this repo ASAP, and let's merge this PR to that branch first.
@binliunls Please update the MONAI version to 1.2.0 directly in this PR.

Thanks.

Comment (Nic-Ma) [reply models/endoscopic_tool_segmentation/configs/metadata.json]: @yiheng-wang-nv I see you already created `dev-1.2` branch, @binliunls let's use it.

Thanks.

File: models/pathology_tumor_detection/docs/README.md
Comment (yiheng-wang-nv) [inline models/pathology_tumor_detection/docs/README.md:95]: Hi @binliunls , could you please also help to remove:
https://github.com/Project-MONAI/model-zoo/blob/dev/ci/bundle_custom_data.py#L30

File: models/endoscopic_tool_segmentation/docs/README.md
Comment (yiheng-wang-nv) [inline models/endoscopic_tool_segmentation/docs/README.md]: should here be: `with fp32 or fp16` according to `<fp32/fp16>`

File: models/endoscopic_tool_segmentation/docs/README.md
Comment (Nic-Ma) [inline models/endoscopic_tool_segmentation/docs/README.md:44]: Please describe the TensorRT version in the doc.

Thanks.

File: models/spleen_ct_segmentation/docs/README.md
Comment (Nic-Ma) [inline models/spleen_ct_segmentation/docs/README.md:52]: Should we use the ONNX workaround here?

Thanks.

File: models/endoscopic_tool_segmentation/docs/README.md
Comment (binliunls) [reply models/endoscopic_tool_segmentation/docs/README.md:44]: I have added the torch-tensorrt version requirment to the CONTRIBUTIONG file and will add the TensorRT requirment. Should we also add these version requirement to every bundle?

File: models/spleen_ct_segmentation/docs/README.md
Comment (binliunls) [reply models/spleen_ct_segmentation/docs/README.md:52]: Ok, I can update this part.

File: models/endoscopic_tool_segmentation/docs/README.md
Comment (Nic-Ma) [reply models/endoscopic_tool_segmentation/docs/README.md:44]: TensorRT requirement is another thing.
Here I mean maybe it's better to describe the **exact** TensorRT version you used to get this benchmark number.

Thanks.

File: models/spleen_ct_segmentation/docs/README.md
Comment (Nic-Ma) [reply models/spleen_ct_segmentation/docs/README.md:52]: And where you plan to put the description for ONNX workaround?

Thanks.

File: models/endoscopic_tool_segmentation/docs/README.md
Comment (binliunls) [reply models/endoscopic_tool_segmentation/docs/README.md:44]: Thanks, I got your point and will update this part.

File: models/spleen_ct_segmentation/docs/README.md
Comment (binliunls) [reply models/spleen_ct_segmentation/docs/README.md:52]: Since currently only the spleen ct segmentation bundle is using onnx workaround, I think we'd better just put it in the README file of spleen ct segmentation. How about this?

Comment (Nic-Ma) [reply models/spleen_ct_segmentation/docs/README.md:52]: But you will update several other bundles with ONNX workaround soon, right?
So I think maybe it's better to describe in the contribution guide directly?

Thanks.

**Comment by yiheng-wang-nv** [comment]
Still cannot make the spleen bundle work with the following settings:
```
GPU: A100 80G
torch_tensorRT version: 1.4.0.dev0
monai version: branch from: https://github.com/Project-MONAI/MONAI/pull/5986  commit cda525ac
bundle commit: 6b8518edf0569508ac30cd7608c7e8bb7348f3b8
```
Export command:
```
python -m monai.bundle trt_export \
--net_id network_def --filepath models/model_trt_fp16.ts \
--ckpt_file models/model.pt --meta_file configs/metadata.json \
--config_file configs/inference.json --precision fp16 \
--dynamic_batchsize "[1, 4, 8]"
```
inference command:
```
python -m monai.bundle run evaluating \
--meta_file configs/metadata.json \
--config_file configs/inference_trt.json \
--logging_file configs/logging.conf
```
Where `inference_trt.json` is modified from `inference.json` according to https://github.com/Project-MONAI/model-zoo/blob/f4aa28e1820e0d65be9855513bf801d0048b76e8/CONTRIBUTING.md#verify-tensorrt-models

**Comment by binliunls** [comment]
> Still cannot make the spleen bundle work with the following settings:
> 
> ```
> GPU: A100 80G
> torch_tensorRT version: 1.4.0.dev0
> monai version: branch from: https://github.com/Project-MONAI/MONAI/pull/5986  commit cda525ac
> bundle commit: 6b8518edf0569508ac30cd7608c7e8bb7348f3b8
> ```
> 
> Export command:
> 
> ```
> python -m monai.bundle trt_export \
> --net_id network_def --filepath models/model_trt_fp16.ts \
> --ckpt_file models/model.pt --meta_file configs/metadata.json \
> --config_file configs/inference.json --precision fp16 \
> --dynamic_batchsize "[1, 4, 8]"
> ```
> 
> inference command:
> 
> ```
> python -m monai.bundle run evaluating \
> --meta_file configs/metadata.json \
> --config_file configs/inference_trt.json \
> --logging_file configs/logging.conf
> ```
> 
> Where `inference_trt.json` is modified from `inference.json` according to https://github.com/Project-MONAI/model-zoo/blob/f4aa28e1820e0d65be9855513bf801d0048b76e8/CONTRIBUTING.md#verify-tensorrt-models

Hi @yiheng-wang-nv ,
      This issue is caused by the `amp` set in the evaluator, which should be set to `false` if using TensorRT models. Because the TensorRT models directly support the fp16/fp32 precision instead of the `amp` mode. I will add this to the CONTRIBUTING file.

**Comment by yiheng-wang-nv** [comment]
Thanks @binliunls , it works when disable amp mode.

File: models/pathology_tumor_detection/configs/metadata.json
Comment (yiheng-wang-nv) [inline models/pathology_tumor_detection/configs/metadata.json:8]: could `torch_tensorrt` version be added into "optional_packages_version" below? Thanks

**Comment by yiheng-wang-nv** [comment]
it seems `FlexibleUNet` (from "endoscopic_tool_segmentation" bundle) has some breaking changes after `monai==1.1.0`, and torchscript test failed on my local test. The error message is:
```
2023-03-21 14:24:06,034 - INFO - --- input summary of monai.bundle.scripts.ckpt_export ---
2023-03-21 14:24:06,035 - INFO - > net_id: 'network_def'
2023-03-21 14:24:06,035 - INFO - > filepath: 'models/endoscopic_tool_segmentation/models/verify_model.ts'
2023-03-21 14:24:06,035 - INFO - > meta_file: 'models/endoscopic_tool_segmentation/configs/metadata.json'
2023-03-21 14:24:06,035 - INFO - > config_file: 'models/endoscopic_tool_segmentation/configs/inference.json'
2023-03-21 14:24:06,035 - INFO - > ckpt_file: 'models/endoscopic_tool_segmentation/models/model.pt'
2023-03-21 14:24:06,035 - INFO - > bundle_root: 'models/endoscopic_tool_segmentation'
2023-03-21 14:24:06,035 - INFO - ---


Traceback (most recent call last):
  File "ci/verify_bundle.py", line 341, in <module>
    verify(bundle, models_path, mode)
  File "ci/verify_bundle.py", line 329, in verify
    verify_torchscript(bundle_path, net_id, config_file)
  File "ci/verify_bundle.py", line 277, in verify_torchscript
    ckpt_export(
  File "/home/venn/Desktop/monai-code/MONAI/monai/bundle/scripts.py", line 988, in ckpt_export
    Checkpoint.load_objects(to_load={key_in_ckpt_: net}, checkpoint=ckpt_file_)
  File "/home/venn/Desktop/monai-code/test_model_zoo/lib/python3.8/site-packages/ignite/handlers/checkpoint.py", line 628, in load_objects
    _load_object(obj, checkpoint_obj)
  File "/home/venn/Desktop/monai-code/test_model_zoo/lib/python3.8/site-packages/ignite/handlers/checkpoint.py", line 620, in _load_object
    obj.load_state_dict(chkpt_obj, strict=is_state_dict_strict)
  File "/home/venn/Desktop/monai-code/test_model_zoo/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1671, in load_state_dict
    raise RuntimeError('Error(s) in loading state_dict for {}:\n\t{}'.format(
RuntimeError: Error(s) in loading state_dict for FlexibleUNet:
        Missing key(s) in state_dict: "decoder.blocks.0.upsample.preconv.weight", "decoder.blocks.0.upsample.preconv.bias", "decoder.blocks.1.upsample.preconv.weight", "decoder.blocks.1.upsample.preconv.bias", "decoder.blocks.2.upsample.preconv.weight", "decoder.blocks.2.upsample.preconv.bias", "decoder.blocks.3.upsample.preconv.weight", "decoder.blocks.3.upsample.preconv.bias".
        size mismatch for decoder.blocks.0.convs.conv_0.conv.weight: copying a param with shape torch.Size([256, 472, 3, 3]) from checkpoint, the shape in current model is torch.Size([256, 296, 3, 3]).
        size mismatch for decoder.blocks.1.convs.conv_0.conv.weight: copying a param with shape torch.Size([128, 304, 3, 3]) from checkpoint, the shape in current model is torch.Size([128, 176, 3, 3]).
        size mismatch for decoder.blocks.2.convs.conv_0.conv.weight: copying a param with shape torch.Size([64, 152, 3, 3]) from checkpoint, the shape in current model is torch.Size([64, 88, 3, 3]).
        size mismatch for decoder.blocks.3.convs.conv_0.conv.weight: copying a param with shape torch.Size([32, 80, 3, 3]) from checkpoint, the shape in current model is torch.Size([32, 48, 3, 3]).
```

**Comment by yiheng-wang-nv** [comment]
The issue seems come from: https://github.com/Project-MONAI/MONAI/pull/5982/files#diff-21a423ef19c6d849ae136606e3094b09efefba90929e80e8c0abbc14b403fe88R312
now the new network contains `preconv` layer and old weights do not have.

**Comment by binliunls** [comment]
> 

Hi @yiheng-wang-nv ,
     I have make a PR https://github.com/Project-MONAI/MONAI/pull/6209 to fix this problem and will update the bundle later.

File: models/pathology_tumor_detection/configs/metadata.json
Comment (yiheng-wang-nv) [reply models/pathology_tumor_detection/configs/metadata.json:8]: Hi @binliunls , it seems `pipenv` cannot find torch tensorrt 1.3.0  I rethought this situation and now I think we can remove it in metadata (we use MONAI docker for tensorrt related tests thus it doesn't matter here)
Could you please help to:
1. remove "torch_tensorrt"
2. modify "monai_version" into `1.2.0rc2`

Comment (Nic-Ma) [reply models/pathology_tumor_detection/configs/metadata.json:8]: Hi @yiheng-wang-nv ,

About the MONAI verison in metadata, I am OK with "1.2.0rc2" or "1.2.0", it's only on `dev-1.2` branch.
Which one do you prefer? I would suggest "1.2.0" because we will upload bundles to NGC for QA testing before MONAI 1.2 release, if no bugs, then we don't need to change anything before really publish the bundles on NGC.

Thanks.

File: models/endoscopic_tool_segmentation/configs/metadata.json
Comment (yiheng-wang-nv) [inline models/endoscopic_tool_segmentation/configs/metadata.json]: ~~should be "1.2.0rc2", as "1.2.0" is not existing~~
I will discuss with Nic first.

File: models/pathology_tumor_detection/configs/metadata.json
Comment (yiheng-wang-nv) [reply models/pathology_tumor_detection/configs/metadata.json:8]: Hi Nic, `1.2.0` will not pass ci test in model zoo, thus I suggested to use `1.2.0rc2`. I think we should use correct version here.

Comment (yiheng-wang-nv) [reply models/pathology_tumor_detection/configs/metadata.json:8]: Otherwise, let me add a workaround to test in dev-1.2, and remove it before merging into dev

File: models/endoscopic_tool_segmentation/docs/README.md
Comment (yiheng-wang-nv) [inline models/endoscopic_tool_segmentation/docs/README.md:106]: met an error on this bundle:
```
Traceback (most recent call last):
  File "/usr/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/usr/local/lib/python3.8/dist-packages/monai/bundle/__main__.py", line 20, in <module>
    fire.Fire()
  File "/usr/local/lib/python3.8/dist-packages/fire/core.py", line 141, in Fire
    component_trace = _Fire(component, args, parsed_flag_args, context, name)
  File "/usr/local/lib/python3.8/dist-packages/fire/core.py", line 466, in _Fire
    component, remaining_args = _CallAndUpdateTrace(
  File "/usr/local/lib/python3.8/dist-packages/fire/core.py", line 681, in _CallAndUpdateTrace
    component = fn(*varargs, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/monai/bundle/scripts.py", line 1137, in trt_export
    _export(
  File "/usr/local/lib/python3.8/dist-packages/monai/bundle/scripts.py", line 907, in _export
    net = converter(model=net, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/monai/networks/utils.py", line 703, in convert_to_trt
    trt_model = torch_tensorrt.compile(
  File "/usr/local/lib/python3.8/dist-packages/torch_tensorrt/_compile.py", line 125, in compile
    return torch_tensorrt.ts.compile(
  File "/usr/local/lib/python3.8/dist-packages/torch_tensorrt/ts/_compiler.py", line 136, in compile
    compiled_cpp_mod = _C.compile_graph(module._c, _parse_compile_spec(spec))
RuntimeError: [Error thrown at core/partitioning/shape_analysis.cpp:176] Expected ivalues_maps.count(input) to be true but got false
Could not find torch::jit::Value* 9351 produced from %9351 : Tensor = aten::_convolution(%x_0.2, %self.decoder.blocks.0.convs.conv_0.conv.weight.3, %self.encoder._conv_stem.bias.33, %11, %11, %11, %9349, %9350, %self.encoder._blocks.0.0.expand_ratio.1, %9349, %9349, %9349, %9349) in lowering graph for mini graph input.
```
command I used:
```
python -m monai.bundle trt_export \
--net_id network_def --filepath models/model_trt_fp32.ts \
--ckpt_file models/model.pt --meta_file configs/metadata.json \
--config_file configs/inference.json --precision fp32
```

File: models/endoscopic_tool_segmentation/configs/metadata.json
Comment (yiheng-wang-nv) [reply models/endoscopic_tool_segmentation/configs/metadata.json]: Hi @binliunls , I just discussed with Nic, and we decided to use "1.2.0rc2" until next release is published.

**Review by yiheng-wang-nv — APPROVED** [review]
Confirmed that three modified bundles support TensorRT according to my local test.
I used one A100 80G GPU to test and the following the export time:
time to test for each bundle (one A100 80G):

- `spleen_ct_segmentation`
    - fp 16: 12 min
    - fp 32: 4 min
- `pathology_tumor_detection`
    - fp 16: 30s
    - fp 32: 16s
- `endoscopic_tool_segmentation`
    - fp 16: 5 min
    - fp 32: 3 min
 
This PR requires the MONAI with commit: c9889ced49b1a9a2bfb828cfed7aa92ca9c74c62, thus it can be merged after we have `monai==1.2.0rc3`

File: models/pathology_tumor_detection/configs/metadata.json
Comment (yiheng-wang-nv) [inline models/pathology_tumor_detection/configs/metadata.json]: mark: to be updated when we have a new version

File: models/spleen_ct_segmentation/docs/README.md
Comment (Nic-Ma) [inline models/spleen_ct_segmentation/docs/README.md:106]: Please add the ONNX conversion command here when your MONAI core PRs are merged.
Pending.

Thanks.

Comment (yiheng-wang-nv) [reply models/spleen_ct_segmentation/docs/README.md:106]: Update: @binliunls will submit a separate PR to address the onnx changes.

**Pull request merged**
